libs/@local/hash-subgraph/src/stdlib/interval.ts

```diff
@@ -0,0 +1,272 @@
+> = [LeftInterval, RightInterval] extends [
+  TimeInterval<infer LeftStartBound, infer LeftEndBound>,
+  TimeInterval<infer RightStartBound, infer RightEndBound>,
+]
+  ? TimeInterval<
+      LeftStartBound | RightStartBound extends LimitedTemporalBound
+        ? LimitedTemporalBound
+        : TemporalBound,
+      LeftEndBound | RightEndBound extends LimitedTemporalBound
+        ? LimitedTemporalBound
+        : TemporalBound
+    >
+  : never;
+
+/**
+ * Returns the intersection (overlapping range) of two given {@link TimeInterval}s, returning `null` if there
+ * isn't any.
+ *
+ * @param {TimeInterval} left
+ * @param {TimeInterval} right
+ */
+export const intervalIntersectionWithInterval = <
+  LeftInterval extends TimeInterval = TimeInterval,
+  RightInterval extends TimeInterval = TimeInterval,
+>(
+  left: LeftInterval,
+  right: RightInterval,
+): IntersectionReturn<LeftInterval, RightInterval> | null =>
+  intervalIntersectionWithIntervalBp<LeftInterval, RightInterval>(
+    left,
+    right,
+  ) as IntersectionReturn<LeftInterval, RightInterval>;
+
+/**
+ * Advanced type to provide stronger type information when using {@link intervalMergeWithInterval}.
+ *
+ * If *both* of the `start` {@link TemporalBound}s are bounded, then the resultant `start` {@link TemporalBound} will be
+ * bounded, same goes for end respectively
+ */
+type MergeReturn<
+  LeftInterval extends TimeInterval,
+  RightInterval extends TimeInterval,
+> = [LeftInterval, RightInterval] extends [
+  TimeInterval<infer LeftStartBound, infer LeftEndBound>,
+  TimeInterval<infer RightStartBound, infer RightEndBound>,
+]
+  ? TimeInterval<
+      LeftStartBound extends LimitedTemporalBound
+        ? RightStartBound extends LimitedTemporalBound
+          ? LimitedTemporalBound
+          : TemporalBound
+        : TemporalBound,
+      LeftEndBound extends LimitedTemporalBound
+        ? RightEndBound extends LimitedTemporalBound
+          ? LimitedTemporalBound
+          : TemporalBound
+        : TemporalBound
+    >
+  : never;
+
+/**
+ * Returns the {@link TimeInterval} which fully spans the space between the `start` {@link TemporalBound}s and
+ * end {@link TemporalBound}s of two provided {@link TimeInterval}s.
+ *
+ * If the intervals do not overlap and are not adjacent, the resultant interval will span _more_ space than that spanned
+ * by the given intervals. _This is different behavior compared to {@link intervalUnionWithInterval}._
+ *
+ * @param {TimeInterval} left
+ * @param {TimeInterval} right
+ */
+export const intervalMergeWithInterval = <
+  LeftInterval extends TimeInterval = TimeInterval,
+  RightInterval extends TimeInterval = TimeInterval,
+>(
+  left: LeftInterval,
+  right: RightInterval,
+): MergeReturn<LeftInterval, RightInterval> =>
+  intervalMergeWithIntervalBp<LeftInterval, RightInterval>(
+    left,
+    right,
+  ) as unknown as MergeReturn<
+    LeftInterval,
+    RightInterval
+  >; /** @todo fix this unknown cast */
```

Comment:
I'm not actually sure why this one is needed, I _think_ the types should be compatible